### PR TITLE
Prevent Quantum Spin Inverter from Teleporting Things into Microwaves

### LINKED ISSUE #29157: Quantum inverter to microwave inventory
## Description
allows you to place yourself into a microwave or any other thing that accepts QSI's

**Reproduction**
1.Place QSI in microwave
2.Use QSI to swap places
3.Now in microwave

This can also be done in a way to remove someone from the round
If they jump into a crate that is QSIed into a microwave they will be trapped forever (unless they have a screwdriver)

### PR DESCRIPTION
**Pull request by Cojoke-dot**
## About the PR

Puts a check on the QSI teleport if the thing that can be teleported would fit into the inventory it would be teleported into.
## Why / Balance

Fixes #29157 and #27397 (as far as I can tell from my testing)
## Technical details

Uses CanInsert to check if both objects being teleported would be able to fit into the inventory of the other object. If not then the teleport is canceled.
## Media

https://github.com/space-wizards/space-station-14/assets/83733158/4f5b05d0-0d78-49fd-9ed5-88eb55915cc1

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- fix: You can no longer teleport objects that should not be in other objects into other objects with the Quantum Spin Inverter

### CODE REVIEW

File: Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs [hunk 1]
```diff
@@ -153,6 +153,14 @@ public void DoTeleport(Entity<SwapTeleporterComponent, TransformComponent> ent)
...
         var otherTeleEnt = GetTeleportingEntity((linkedEnt, Transform(linkedEnt)));
 
+        if (_container.TryGetOuterContainer(teleEnt, Transform(teleEnt), out var othercontainer) &&
+            !_container.CanInsert(teleEnt, othercontainer))
+            return;
+
+        if (_container.TryGetOuterContainer(otherTeleEnt, Transform(otherTeleEnt), out var container) &&
+            !_container.CanInsert(otherTeleEnt, container))
+            return;
+
```
Comment (metalgearsloth) [inline Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs:163]: GetTeleportingEntity should be updated for this instead to be a TryGet.

**Comment by Cojoke-dot** [comment]
So... it turns out the only reason the player can't get into microwaves and other players' hands is not having the item component, something that CanInsert does not check for. That means CanInsert has no meaningful effect here. It seems to work the same without the CanInsert with the testing I did.

File: Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using Content.Shared.Teleportation.Components;
 using Content.Shared.Verbs;
 using Content.Shared.Whitelist;
+using Linguini.Syntax.Ast;
```
Comment (metalgearsloth) [inline Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs:8]: This doesn't look like it belongs.

File: Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs [hunk 3]
```diff
@@ -184,20 +194,20 @@ public void DestroyLink(Entity<SwapTeleporterComponent?> ent, EntityUid? user)
...
         if (!_xformQuery.TryGetComponent(parent, out var parentXform) || parentXform.Anchored)
-            return ent;
+            return (ent, container);
 
         if (!TryComp<PhysicsComponent>(parent, out var body) || body.BodyType == BodyType.Static)
-            return ent;
+            return (ent, container);
 
         return GetTeleportingEntity((parent, parentXform));
     }
```
Comment (metalgearsloth) [inline Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs:213]: This looks like it doesn't check caninsert anymore?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs (hunk 3 above)
Comment (Cojoke-dot) [reply Content.Shared/Teleportation/Systems/SwapTeleporterSystem.cs:213]: CanInsert does not do a check that would prevent a player from getting teleported into a microwave or another player's hand. 
image

https://github.com/space-wizards/space-station-14/assets/83733158/86110f4b-c2ee-4b4d-88d6-6f536f3c757d

**Review by metalgearsloth — APPROVED** [review]

**Comment by slarticodefast** [comment]
Looks like this PR broke microwave construction.

**Pull request merged**